extractor/docker/docker.go

```diff
@@ -257,29 +311,138 @@ func (d Extractor) extractLayerFiles(layerCh chan layer, errCh chan error, ctx c
 	return nil
 }
 
-func (d Extractor) extractLayerWorker(dig digest.Digest, r *registry.Registry, ctx context.Context, image registry.Image, errCh chan error, layerCh chan layer) {
-	var rc io.Reader
-	// Use cache
-	rc = d.Cache.Get(string(dig))
-	if rc == nil {
-		// Download the layer.
-		layerRC, err := r.DownloadLayer(ctx, image.Path, dig)
+func (d Extractor) extractLayerWorker(dig digest.Digest, r *registry.Registry, ctx context.Context, image registry.Image, errCh chan error, layerCh chan layer, filenames []string) {
+	var tarContent bytes.Buffer
+	var cacheContent []byte
+	found, _ := d.Cache.Get(kvtypes.GetItemInput{
+		BucketName: "layertars",
+		Key:        string(dig),
+		Value:      &cacheContent,
+	})
+
+	if found {
+		b, err := extractTarContent(cacheContent)
+		if err != nil || len(b) <= 0 {
+			found = false
+		}
+
+		_, err = tarContent.Write(b)
+		if err != nil {
+			found = false
+		}
+	}
+
+	if !found {
+		rc, err := r.DownloadLayer(ctx, image.Path, dig)
 		if err != nil {
 			errCh <- xerrors.Errorf("failed to download the layer(%s): %w", dig, err)
 			return
 		}
+		defer rc.Close()
 
-		rc, err = d.Cache.Set(string(dig), layerRC)
+		// read the incoming gzip from the layer
+		gzipReader, err := gzip.NewReader(rc)
 		if err != nil {
-			log.Print(err)
+			errCh <- xerrors.Errorf("could not init gzip reader: %w", err)
+			return
+		}
+		defer gzipReader.Close()
+
+		tarReader := tar.NewReader(io.TeeReader(gzipReader, &tarContent))
```

Comment:
`cacheBuf` should be returned as `tarContent`, isn't it? If we use `io.TeeReader`, it seems that the raw tar content will be returned.

Reply:
Yes but that's what we're after right? After passing the tar to content the Extract method will be called which expects a tar file https://github.com/aquasecurity/fanal/blob/5f894c996176f0bfb2faa6f5be81338a7f179e10/extractor/docker/docker.go#L531-L535

Maybe I misunderstood what you meant to say. Do you have a diff or a test case in mind?

Reply:
I see. Thanks that makes sense. I have fixed this here: https://github.com/aquasecurity/fanal/pull/51/commits/745ac0cbcb8105744323beef89cddf7d5e9449ef